### PR TITLE
Benchmarking Demo App

### PR DESCRIPTION
**Pull request by jbedichekTT**
Added a streamlit UI web application with performance benchmarking for 5 models:
1. BERT
2. MNIST
3. CodeGen
4. ResNet18
5. GPT2

Instructions for how to use it are in the application itself, to run the demo:
1. pip install -r requirements-demo.txt
2. streamlit run demo.py
3. Command click on the local URL to open the UI in a web broswer.

Some of the options/features:
1. Measures tokens/sec for BERT, GPT2, CodeGen
2. Optionally compares TTNN to CPU speed, computes relative speedup 
3. Can specify num_iterations (i.e 5), measures the latency of the last run.
4. No need to prompt any model, default values already set.

It may take up to a 5 minutes to generate an output, depending on parameters.

Here's a sample output showing BERT performance:
demo

### CODE REVIEW

File: requirements-demo.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+streamlit
```
Comment (jmalone-tt) [inline requirements-demo.txt:1]: Add `-r requirements.txt` to top. Use case here would be someone wanting to run the demo, they should just be able to run `pip install -r requirements-demo.txt` and be set. That line will also install all other dependencies they need.

File: demo.py [hunk 2]
```diff
@@ -0,0 +1,634 @@
...
+        if use_ttnn:
+            ttnn.close_device(device)
+            print("TTNN device closed.")
+        return label, confidence_score, inference_time
+
+    def stream_response(response):
+        for char in response:
+            yield char
+
+    def main():
```
Comment (jmalone-tt) [inline demo.py:310]: main is pretty long. Breaking it into smaller functions could help readability. It might be good to have different files for each of the models

File: demo.py [hunk 3]
```diff
@@ -0,0 +1,634 @@
...
+            if len(words) > 200:
+                output_str = " ".join(words[:200]) + " [Truncated...]"
+            return result, output_str
+        finally:
+            sys.stdout = old_stdout
+
+    return wrapper
+
+
+with torch.no_grad():
```
Comment (jmalone-tt) [inline demo.py:75]: Should this line be here?

File: demo.py [hunk 4]
```diff
@@ -0,0 +1,634 @@
...
+            model = torch.compile(model, backend=torch_ttnn.backend, options=option)
+            print("Initial TTNN generation...")
+            for i in range(iterations - 1):
+                _ = model.generate(
+                    **inputs,
+                    max_length=max_length,
+                    num_return_sequences=num_return_sequences,
+                    do_sample=True,
+                )
+        print(f"Generating text for {iterations} iterations...")
```
Comment (jmalone-tt) [inline demo.py:198]: Would help maintainability to factor this out into a reusable function. Each model wants this behavior, so it's better to define it once if possible. Same feedback for any other repeated actions in each model's function

File: demo.py [hunk 5]
```diff
@@ -0,0 +1,634 @@
...
+                "Code Generation with CodeGen",
+            ],
+        )
+
+        st.subheader(f"About {task}")
+        if task == "Question Answering with BERT":
+            st.info(
+                """
+                **BERT**: Fine tuned for finding the answers to questions in a given context (SQuAD 2.0 https://rajpurkar.github.io/SQuAD-explorer/).
+                **How to Use**: Set context and question, set iterations, then click "Run".
```
Comment (jmalone-tt) [inline demo.py:329]: Thanks for adding notes on how to use!

File: demo.py [hunk 6]
```diff
@@ -0,0 +1,634 @@
...
+                        if test_cpu:
+                            (cpu_logits, cpu_label, cpu_confidence, cpu_time), cpu_output = classify_mnist(
+                                use_ttnn=False, iterations=iterations
+                            )
+                        full_response = f"Predicted Digit: {ttnn_label} (Confidence: {ttnn_confidence:.2f}%)"
+                        message_placeholder.markdown(full_response)
+                        st.write("**Logits (TTNN):**")
+                        st.write(ttnn_logits.numpy())
+                        terminal_output_value = f"TTNN:\n{ttnn_output}"
+                        if test_cpu:
```
Comment (jmalone-tt) [inline demo.py:595]: Is there a reason to re-check test_cpu several times?

**Review by jmalone-tt — CHANGES_REQUESTED** [review]
Left a couple comments, thanks for working on this!

File: requirements-demo.txt (hunk 1 above)
Comment (jbedichekTT) [reply requirements-demo.txt:1]: Will do

**Comment by ayerofieiev-tt** [comment]
Can you please add some visuals to the PR description?

File: demo.py (hunk 3 above)
Comment (jbedichekTT) [reply demo.py:75]: No, it's redundant, will remove.

File: demo.py (hunk 6 above)
Comment (jbedichekTT) [reply demo.py:595]: I think it needs to be that way because its value bound to the checkbox, if someone unchecks it, it shouldn't display the CPU performance.

File: demo_models/demo.md [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+<p align="center">
+  <img src="img/streamlit.png" alt="config" width="500"/>
+</p>
+
+## How To Use The App
+**Select Model To Test**:
+There are five models available to run, choose the model and task from the drop down menu near the top of the page.
+
+**Configure The Test**:
+The test can be configured with a prompt or image to test, default values are already set for convenience. The test can also be configured with an option to measure CPU performance for comparison.  The test returns a latency value, and a tokens/sec measurement for language models. The latency is the time it takes to run a forward pass for BERT, MNSIT, or ResNet. For GPT2 and CodeGen, the latency is the time it takes to run a full sequence generation.  This is with a configurable input and sequence generation length. These generative language models, along with BERT, also return a tokens/sec measurement.
```
Comment (jmalone-tt) [inline demo_models/demo.md:53]: MNSIT -> MNIST

File: demo_models/demo.md [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+**Run The Test**:
+The test will display a graph of execution time vs warmup iteration, it will return an execution time, terminal output, relative CPU speedup, and for text models, tokens/sec. 
+
+<p align="center">
+  <img src="img/output.png" alt="config" width="500"/>
+</p>
+
+## Model Information
+**Question Answering With BERT**:
+The version of BERT used in this demo is the "BERT-Large" variant, fine tuned on the [SQuAD 2.0 dataset](https://rajpurkar.github.io/SQuAD-explorer/).  This model is designed to answer questions by processing a question text and a context text together as one input. This context is a text which is expected to contain the exact answer to the question. The model finds a subset of this text and returns it as an answer to the question. It works by assinging a probability to each token in the context string of being the start and end of the returned answer. The amount of tensors getting processed by the model scale with the length of the question and context.
```
Comment (jmalone-tt) [inline demo_models/demo.md:68]: assinging

File: demo_models/demo.md [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+
+In the actual test in this script, the sequences are padded (or truncated) to a prespecified length, configurable by adjusting the "sequence length" value below the question box.  The padding represent unlearned tokens from the model which are processed by its components, but are masked to not influence the output. This padding/truncation ensures a consistent tokens/sec measurement which doesn't depend on the specific question and context being processed by the model, since padding tokens are included in the computations of the model in the same way as regular tokens.  This means, in this script, tokens/sec is measured as ```sequence_length / execution_time```, ensuring consistent outputs.
```
Comment (jmalone-tt) [inline demo_models/demo.md:70]: The padding represent unlearned tokens

File: demo_models/demo.md [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+
+
+
+**GPT-2**
+This is a generative language model form open AI, specifically the gpt2-small variant with 125 million parameters.  It is a decoder only transformer which generates outputs autoregressively.  The input is padded (or truncated) to a configurable number of tokens just like BERT, by setting ```Input Sequence Length``` below the prompt. The generation length is also configurable with the ```New Tokens To Generate``` input.  This padding and truncation ensures consistent latency measurements for different inputs.
```
Comment (jmalone-tt) [inline demo_models/demo.md:75]: form -> from

File: demo_models/demo.md [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+  <img src="img/codegen.png" alt="config" width="500"/>
+</p>
+
+
+**ResNet18**
+This is an image classification model incorporating convolutions and normalization ops heavily.  It can process 32x32x3 color images. Only execution time is measured.
+
+## Troubleshooting
+**Compilation Errors**
+ResNet18 has been known to throw compilation errors, specifically after ```ConstantFoldingPass```.  This is due to issues in model specific integration for TTNN.
```
Comment (jmalone-tt) [inline demo_models/demo.md:98]: Is this still an issue? I would expect those errors to be fixed now

**Review by jmalone-tt — CHANGES_REQUESTED** [review]
A couple more changes, but the README looks really good overall. Please run spellcheck too in case I missed any.

I would also request all of the files related to the demo be put into a top level demo directory similar to the `tests` directory containing all files related to tests.

Thanks for working on this - I'm excited to get it merged!

File: demo_models/demo.md (hunk 11 above)
Comment (jbedichekTT) [reply demo_models/demo.md:98]: ResNet18 still is not compiling using my setup, and I'm using a relatively recent version of metal.

**Review by jmalone-tt — APPROVED** [review]
